File: backend/app/routers/downloads.py
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.db.database import get_db
from app.schemas.downloads import (
    DownloadTaskDeleteResponse,
    DownloadTaskListResponse,
    DownloadTaskMessageResponse,
)
from app.services.download_service import (
    create_download_task,
    delete_download_task,
    list_download_tasks,
    refresh_download_task,
)

router = APIRouter(tags=["下载队列"])
DbSession = Annotated[Session, Depends(get_db)]


@router.get("/api/downloads", response_model=DownloadTaskListResponse)
def read_downloads(db: DbSession) -> DownloadTaskListResponse:
    return DownloadTaskListResponse(
        message="下载任务列表获取成功",
        downloads=list_download_tasks(db),
    )


@router.post(
    "/api/source-items/{item_id}/download",
    response_model=DownloadTaskMessageResponse,
    status_code=status.HTTP_201_CREATED,
)
def add_download(item_id: int, db: DbSession) -> DownloadTaskMessageResponse:
    try:
        download = create_download_task(db, item_id)
    except LookupError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc
    return DownloadTaskMessageResponse(message="下载任务已创建", download=download)


@router.post("/api/downloads/{download_id}/refresh", response_model=DownloadTaskMessageResponse)
def refresh_download(download_id: int, db: DbSession) -> DownloadTaskMessageResponse:
    try:
        download = refresh_download_task(db, download_id)
    except LookupError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc
    return DownloadTaskMessageResponse(message="下载任务状态已刷新", download=download)


@router.delete("/api/downloads/{download_id}", response_model=DownloadTaskDeleteResponse)
def remove_download(download_id: int, db: DbSession) -> DownloadTaskDeleteResponse:
    try:
        delete_download_task(db, download_id)
    except LookupError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    return DownloadTaskDeleteResponse(message="下载任务记录已删除")

File: frontend/src/api/downloads.ts
import type {
  DownloadTaskDeleteResponse,
  DownloadTaskListResponse,
  DownloadTaskMessageResponse
} from '../types/downloads'

const API_BASE = '/api'

async function request<T>(path: string, options?: RequestInit): Promise<T> {
  const response = await fetch(`${API_BASE}${path}`, {
    headers: {
      'Content-Type': 'application/json',
      ...(options?.headers ?? {})
    },
    ...options
  })

  if (!response.ok) {
    let message = '请求失败，请稍后重试'
    try {
      const data = (await response.json()) as { message?: string; detail?: string }
      message = data.message ?? data.detail ?? message
    } catch {
      message = '请求失败，请检查后端服务是否正常'
    }
    throw new Error(message)
  }

  return (await response.json()) as T
}

export function getDownloads(): Promise<DownloadTaskListResponse> {
  return request<DownloadTaskListResponse>('/downloads')
}

export function createDownloadTask(itemId: number): Promise<DownloadTaskMessageResponse> {
  return request<DownloadTaskMessageResponse>(`/source-items/${itemId}/download`, {
    method: 'POST'
  })
}

export function refreshDownloadTask(downloadId: number): Promise<DownloadTaskMessageResponse> {
  return request<DownloadTaskMessageResponse>(`/downloads/${downloadId}/refresh`, {
    method: 'POST'
  })
}

export function deleteDownloadTask(downloadId: number): Promise<DownloadTaskDeleteResponse> {
  return request<DownloadTaskDeleteResponse>(`/downloads/${downloadId}`, {
    method: 'DELETE'
  })
}
